[clang] Fix assertion failure with explicit(bool) in pre-C++11 modes

jongmyeong-choi opened the pull request:
Allow CCEKind::ExplicitBool in BuildConvertedConstantExpression for pre-C++11 contexts, similar to the existing TempArgStrict exception. This enables explicit(bool) to work as a C++20 extension in earlier language modes without triggering assertion failures.

Fixes #152729

Commit:
[clang] Fix assertion failure with explicit(bool) in pre-C++11 modes

Allow CCEKind::ExplicitBool in BuildConvertedConstantExpression for
pre-C++11 contexts, similar to the existing TempArgStrict exception.
This enables explicit(bool) to work as a C++20 extension in earlier
language modes without triggering assertion failures.

Fixes #152729

AaronBallman reviewed (commented):
Thank you for the fix!

mizvekov reviewed (approved):
LGTM sans small nit.

FYI the change is correct as far as if we want to allow this as an extension.

But we otherwise don't produce an extension warning here.

Usually when we allow something as an extension to earlier standards, this warning is added, which users can also either ignore it or turn it into an error, and is helpful for people who want strict standard conformance (ie want to make sure this code will compile in other compilers).

Commit:
Address review feedback

  - Add [[maybe_unused]] to suppress unused variable warning
  - Add C++98 test case
  - Fix missing newline

shafik reviewed (commented):
This need a release note in `clang/docs/ReleaseNotes.rst`. This goes back to clang-10.

Commit:
Add release note for explicit(bool) assertion fix

erichkeane commented:
Looks like the documentation failure is related to some OMP documentation, so just merging this.
